topics/reference_guide/custom_language_support/language_server_protocol.md

```diff
@@ -0,0 +1,175 @@
+
+For more complex cases, the plugin may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.
+
+To implement a minimal LSP plugin, perform the following steps:
+- Within the `LspServerSupportProvider.fileOpened()` method, spin up the relevant LSP server descriptor, which can decide if the given file is supported by using the `LspServerDescriptor.isSupportedFile()` check method.
+- Tell how to start the server by implementing `LspServerDescriptor.createCommandLine()`.
+
+## Customization
+
+- To fine-tune or disable the implementation of LSP-based features, the plugins may override the corresponding properties of the `LspServerDescriptor` class.
+  See the properties documentation for more details:
+  - `lspGoToDefinitionSupport`
+  - `lspCompletionSupport`
+  - `lspDiagnosticsSupport`
+  - `lspCodeActionsSupport`
+  - `lspCommandsSupport`
+- To handle custom (undocumented) requests and notifications from the LSP server, override `LspServerDescriptor.createLsp4jClient`.
+- To send custom (undocumented) requests and notifications to the LSP server, override `LspServerDescriptor.lsp4jServerClass` and implement the `LspClientNotification` and/or `LspRequest` classes.
+
+See bundled LSP API source code and its documentation for more information.
+
+## Troubleshooting
+
+All the IDE and LSP server communication logs are passed to the IDE log file.
+
+To include them for preview, add the following entry to the <control>Help | Diagnostic Tools | Debug Log Settings…</control> configuration dialog:
+
+```
+#com.intellij.platform.lsp
+```
+
+For more information, see the [](ide_infrastructure.md#logging) section.
+
+## Limitations
+
+- The current LSP API implementation assumes that the IDE <-> LSP server communication channel is `stdio`.
+- The IDE doesn't send [workspace/didChangeWatchedFiles](https://microsoft.github.io/language-server-protocol/specification/#workspace_didChangeWatchedFiles) notifications to the server.
+
+## Integration Overview
+
+Integrating the Language Server Protocol (LSP) into a plugin for IntelliJ-based IDEs involves a trade-off between simple and fast language support and a complex custom language support plugin with IDE capabilities.
+
+When considering the LSP-based approach, it is important to assess the following criteria for providing a Language Server to end users:
+
+- OS dependency of the Language Server.
+- Availability of the latest version online.
+- Compatibility with breaking changes between versions.
+- Feasibility of requesting the user to provide the Language Server binary path.
+
+If you encounter any issues or need assistance, please provide feedback by reaching out to us through the `#intellij-platform` channel in our [JetBrains Platform Slack](https://plugins.jetbrains.com/slack/) workspace or by submitting an issue in [YouTrack](https://youtrack.jetbrains.com/newIssue?project=IDEA&c=Subsystem%20Core.%20Platform%20API).
```

Comment:
do not duplicate content from https://plugins.jetbrains.com/docs/intellij/getting-help.html#problems-with-code-support-issues

Reply:
add section "sample plugins" with handpicked good samples
and link to https://plugins.jetbrains.com/intellij-platform-explorer/extensions?extensions=com.intellij.platform.lsp.serverSupportProvider for full list of existing LSP plugins